File: Interview Problems/Parking Lot/Java Code/service/AdminService.java
package service;

import domain.Floor;
import domain.ParkingSlot;
import domain.PricingRule;
import domain.Vehicle;
import repository.FloorRepository;
import repository.PricingRuleRepository;
import repository.SlotRepository;

import java.util.List;
import java.util.Map;
import java.util.Optional;

public class AdminService {

    private final FloorRepository floorRepository;
    private final SlotRepository slotRepository;
    private final PricingRuleRepository pricingRuleRepository;

    public AdminService(FloorRepository floorRepository,
                        SlotRepository slotRepository,
                        PricingRuleRepository pricingRuleRepository) {
        this.floorRepository = floorRepository;
        this.slotRepository = slotRepository;
        this.pricingRuleRepository = pricingRuleRepository;
        System.out.println("[SERVICE] AdminService initialized");
    }

    public Floor addFloor(int floorNumber) {
        System.out.println("[ADMIN] Adding floor: " + floorNumber);
        if (floorRepository.existsByFloorNumber(floorNumber)) {
            throw new IllegalArgumentException("Floor " + floorNumber + " already exists");
        }
        Floor floor = new Floor(floorNumber);
        floorRepository.save(floor);
        System.out.println("[ADMIN] Floor " + floorNumber + " added successfully");
        return floor;
    }

    public ParkingSlot addSlot(int floorNumber, Vehicle.VehicleType slotType) {
        System.out.println("[ADMIN] Adding slot of type " + slotType + " to floor " + floorNumber);

        Floor floor = floorRepository.findByFloorNumber(floorNumber)
                .orElseThrow(() -> new IllegalArgumentException("Floor not found: " + floorNumber));

        ParkingSlot slot = new ParkingSlot(slotType, floorNumber);
        slotRepository.save(slot);
        floor.addSlot(slot);

        System.out.println("[ADMIN] Slot added: " + slot.getId() + " on floor " + floorNumber);
        return slot;
    }

    public void updatePricing(Vehicle.VehicleType vehicleType, double ratePerHour, double flatRate) {
        System.out.println("[ADMIN] Updating pricing for: " + vehicleType);
        Optional<PricingRule> existing = pricingRuleRepository.findByVehicleType(vehicleType);
        if (existing.isPresent()) {
            existing.get().updateHourlyRate(ratePerHour);
            existing.get().updateFlatRate(flatRate);
            pricingRuleRepository.update(existing.get());
        } else {
            PricingRule rule = new PricingRule(vehicleType, ratePerHour, flatRate, PricingRule.RuleType.HOURLY);
            pricingRuleRepository.save(rule);
        }
        System.out.println("[ADMIN] Pricing updated for " + vehicleType
                + " | Hourly: ₹" + ratePerHour + " | Flat: ₹" + flatRate);
    }

    public void updateFlatPricing(Vehicle.VehicleType vehicleType, double flatRate) {
        System.out.println("[ADMIN] Updating flat rate for: " + vehicleType + " to ₹" + flatRate);
        PricingRule rule = pricingRuleRepository.findByVehicleType(vehicleType)
                .orElseThrow(() -> new IllegalArgumentException("No pricing rule for: " + vehicleType));
        rule.updateFlatRate(flatRate);
        pricingRuleRepository.update(rule);
    }

    public void updateHourlyPricing(Vehicle.VehicleType vehicleType, double ratePerHour) {
        System.out.println("[ADMIN] Updating hourly rate for: " + vehicleType + " to ₹" + ratePerHour);
        PricingRule rule = pricingRuleRepository.findByVehicleType(vehicleType)
                .orElseThrow(() -> new IllegalArgumentException("No pricing rule for: " + vehicleType));
        rule.updateHourlyRate(ratePerHour);
        pricingRuleRepository.update(rule);
    }

    public void viewParkingStatus() {
        System.out.println("\n========== PARKING STATUS ==========");
        Map<Integer, Floor> floors = floorRepository.findAll();
        if (floors.isEmpty()) {
            System.out.println("No floors configured.");
        }
        for (Map.Entry<Integer, Floor> entry : floors.entrySet()) {
            Floor floor = entry.getValue();
            List<ParkingSlot> slots = floor.getSlots();
            long occupied = slots.stream().filter(ParkingSlot::isOccupied).count();
            System.out.println("Floor " + floor.getFloorNumber()
                    + " | Total slots: " + slots.size()
                    + " | Occupied: " + occupied
                    + " | Available: " + (slots.size() - occupied));
        }
        System.out.println("=====================================\n");
    }
}

File: Interview Problems/Parking Lot/Python Code/service/admin_service.py
from typing import Optional, Dict
from domain.floor import Floor
from domain.parking_slot import ParkingSlot
from domain.vehicle import VehicleType
from domain.pricing_rule import PricingRule, RuleType
from repository.floor_repository import FloorRepository
from repository.slot_repository import SlotRepository
from repository.pricing_rule_repository import PricingRuleRepository

class AdminService:
    def __init__(self, floor_repo: FloorRepository, slot_repo: SlotRepository, pricing_repo: PricingRuleRepository):
        self.floor_repository = floor_repo
        self.slot_repository = slot_repo
        self.pricing_rule_repository = pricing_repo
        print("[SERVICE] AdminService initialized")

    def add_floor(self, floor_number: int) -> Floor:
        print(f"[ADMIN] Adding floor: {floor_number}")
        if self.floor_repository.exists_by_floor_number(floor_number):
            raise ValueError(f"Floor {floor_number} already exists")
        floor = Floor(floor_number)
        self.floor_repository.save(floor)
        print(f"[ADMIN] Floor {floor_number} added successfully")
        return floor

    def add_slot(self, floor_number: int, slot_type: VehicleType) -> ParkingSlot:
        print(f"[ADMIN] Adding slot of type {slot_type.name} to floor {floor_number}")
        floor = self.floor_repository.find_by_floor_number(floor_number)
        if not floor:
            raise ValueError(f"Floor not found: {floor_number}")

        slot = ParkingSlot(slot_type, floor_number)
        self.slot_repository.save(slot)
        floor.add_slot(slot)

        print(f"[ADMIN] Slot added: {slot.id} on floor {floor_number}")
        return slot

    def update_pricing(self, vehicle_type: VehicleType, rate_per_hour: float, flat_rate: float):
        print(f"[ADMIN] Updating pricing for: {vehicle_type.name}")
        existing = self.pricing_rule_repository.find_by_vehicle_type(vehicle_type)
        if existing:
            existing.update_hourly_rate(rate_per_hour)
            existing.update_flat_rate(flat_rate)
            self.pricing_rule_repository.update(existing)
        else:
            rule = PricingRule(vehicle_type, rate_per_hour, flat_rate, RuleType.HOURLY)
            self.pricing_rule_repository.save(rule)
        print(f"[ADMIN] Pricing updated for {vehicle_type.name} | Hourly: ₹{rate_per_hour} | Flat: ₹{flat_rate}")

    def update_flat_pricing(self, vehicle_type: VehicleType, flat_rate: float):
        print(f"[ADMIN] Updating flat rate for: {vehicle_type.name} to ₹{flat_rate}")
        rule = self.pricing_rule_repository.find_by_vehicle_type(vehicle_type)
        if not rule:
            raise ValueError(f"No pricing rule for: {vehicle_type.name}")
        rule.update_flat_rate(flat_rate)
        self.pricing_rule_repository.update(rule)

    def update_hourly_pricing(self, vehicle_type: VehicleType, rate_per_hour: float):
        print(f"[ADMIN] Updating hourly rate for: {vehicle_type.name} to ₹{rate_per_hour}")
        rule = self.pricing_rule_repository.find_by_vehicle_type(vehicle_type)
        if not rule:
            raise ValueError(f"No pricing rule for: {vehicle_type.name}")
        rule.update_hourly_rate(rate_per_hour)
        self.pricing_rule_repository.update(rule)

    def view_parking_status(self):
        print("\n========== PARKING STATUS ==========")
        floors = self.floor_repository.find_all()
        if not floors:
            print("No floors configured.")
        for floor_num, floor in floors.items():
            slots = floor.slots
            occupied = sum(1 for s in slots if s.is_occupied)
            print(f"Floor {floor.floor_number} | Total slots: {len(slots)} | Occupied: {occupied} | Available: {len(slots) - occupied}")
        print("=====================================\n")
